src/dataflow-types/src/types.rs

```diff
@@ -856,35 +936,24 @@ pub mod sources {
                     };
                     keyed.concat(metadata_desc)
                 }
-                SourceEnvelope::Debezium(..) => {
-                    // TODO [btv] - Get rid of this, when we can. Right now source processing is still
-                    // not fully orthogonal, so we have some special logic in the debezium processor
-                    // that only passes on the first two columns ("before" and "after"), and uses the
-                    // information in the other ones itself
-                    let mut diter = value_desc.into_iter();
-                    let mut key = diter
-                        .next()
-                        .ok_or_else(|| anyhow::anyhow!("Spec does not contain key"))?;
-                    match &mut key.1.scalar_type {
+                SourceEnvelope::Debezium(envelope) => {
+                    match &value_desc.typ().column_types[envelope.after_idx].scalar_type {
                         ScalarType::Record { fields, .. } => {
-                            fields.extend(metadata_desc.iter().map(|(k, v)| (k.clone(), v.clone())))
+                            let mut desc = RelationDesc::from_names_and_types(fields.clone());
+                            if let Some(key) = key_desc.map(|k| match_key_indices(&k, &desc)) {
+                                desc = desc.with_key(key?);
+                            }
+
+                            match envelope.mode {
+                                DebeziumMode::Upsert => desc.concat(metadata_desc),
```

Comment:
The envelope processing stage is the moment that a timely stream is converted into a differential collection. This means that we need to be able to issue retractions for rows that change in the future. The crux of the issue is that the metadata that are present in the kafka message that encodes an insertion are different than the metadata of the message that encodes the retraction. For example the offset will be different (since they are two distinct messages).

Upsert is the only mode where we have memory of previous rows emitted and therefore we can recover the metadata we emitted in the past to retract them. All other debezium modes are stateless (at least as far as metadata go) and so cannot have metadata appended to them.